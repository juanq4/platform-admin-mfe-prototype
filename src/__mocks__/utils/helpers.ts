import { isEmpty } from "@q4/nimbus-ui";

export function mockWindowMethods(): Window {
  return Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: jest.fn().mockImplementation((query) => ({
      matches: false,
      media: query,
      onchange: null,
      addListener: jest.fn(), // Deprecated
      removeListener: jest.fn(), // Deprecated
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    })),
  });
}

export function mockClipboardWrite(): void {
  const orginalClipboard = { ...navigator.clipboard };

  beforeAll(() => {
    if (!isEmpty(navigator.clipboard)) return;

    Object.defineProperty(navigator, "clipboard", {
      value: {
        writeText: () => Promise.resolve(),
      },
    });
  });

  afterAll(() => {
    Object.assign(navigator.clipboard, {
      ...orginalClipboard,
    });
  });
}
