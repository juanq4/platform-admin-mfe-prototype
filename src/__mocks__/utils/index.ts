export * from "./helpers";
export * from "./permissions";
export * from "./wrappers";
