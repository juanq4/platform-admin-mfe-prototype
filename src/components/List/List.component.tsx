import type { Key } from "react";
import { Children } from "react";
import type { ListProps } from "./List.definition";
import { Item, Header, Container } from "./List.style";

const List = <T,>(props: ListProps<T>): JSX.Element => {
  const { header, dataSource, renderItem, children, rowKey, hasKeyLine, footer, ...rest } = props;

  const listItemsKeys: { [index: number]: Key } = {};

  const renderInnerItem = (item: T, index: number) => {
    if (!renderItem) return null;

    let key;

    if (typeof rowKey === "function") {
      key = rowKey(item);
    } else if (rowKey) {
      key = item[rowKey];
    } else {
      key = item.toString();
    }

    listItemsKeys[index] = key as Key;

    return renderItem(item, index);
  };

  const items = (dataSource || []).map((item, index) => renderInnerItem(item, index));
  const childrenList = Children.map(items, (child, index) => (
    <Item hasKeyLine={hasKeyLine} key={listItemsKeys[index]}>
      {child}
    </Item>
  ));

  return (
    <Container {...rest}>
      {header && <Header>{header}</Header>}
      {childrenList}
      {children}
      {footer && <div>{footer}</div>}
    </Container>
  );
};

export { List };
