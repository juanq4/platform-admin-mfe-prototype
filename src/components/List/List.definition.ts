import type { ReactNode } from "react";

export interface ListProps<T> extends React.ComponentPropsWithoutRef<"ul"> {
  header?: ReactNode;
  dataSource?: T[];
  rowKey?: ((item: T) => React.Key) | keyof T;
  hasKeyLine?: boolean;
  footer?: React.ReactNode;
  renderItem?: (item: T, index: number) => React.ReactNode;
}
