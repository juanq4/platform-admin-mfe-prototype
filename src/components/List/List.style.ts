import styled from "@emotion/styled";
import type { ListProps } from "./List.definition";

export const Header = styled.div`
  height: 28px;
  width: 100%;
  padding: 4px 0px;
  font-family: ${({ theme }) => theme.typography.openSans};
  font-style: normal;
  font-weight: normal;
  font-size: 14px;
  line-height: 20px;
  color: ${({ theme }) => theme.colors.neutral500};
`;

export const Container = styled.ul`
  list-style: none;
  margin: 0 0 20px 0;
  padding: 0;
`;

export const Item = styled.li<Pick<ListProps<"">, "hasKeyLine">>`
  height: auto;
  font-size: 14px;
  line-height: 16px;
  align-items: center;
  box-shadow: inset 0px -1px 0px ${(props) => (props.hasKeyLine ? props.theme.colors.neutral100 : "inherit")};
`;
