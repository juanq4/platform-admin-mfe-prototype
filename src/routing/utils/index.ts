export * from "./api";
export * from "./s3checks";
export * from "./versionManager";
