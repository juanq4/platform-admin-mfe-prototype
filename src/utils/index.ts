export * from "./account";
export * from "./api";
export * from "./entitlement";
export * from "./error";
export * from "./feature";
export * from "./logout";
export * from "./organization";
export * from "./permission";
export * from "./role";
export * from "./route";
export * from "./subscription";
export * from "./user";
export * from "./environment";
