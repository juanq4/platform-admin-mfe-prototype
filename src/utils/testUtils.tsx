/* eslint-disable import/export */
import { MockedProvider as ApolloMockedProvider } from "@apollo/client/testing";
import { ThemeProvider } from "@emotion/react";
import { StyleGuide, ConfigProvider as NimbusConfig } from "@q4/nimbus-ui";
import type { RenderOptions, RenderResult } from "@testing-library/react";
import { render } from "@testing-library/react";
import theme from "../styles/theme";

// https://testing-library.com/docs/react-testing-library/setup/
const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "queries">,
  styleGuide = StyleGuide.V2,
): RenderResult =>
  render(ui, {
    // Common middleware to all tests
    wrapper: ({ children }) => (
      <NimbusConfig styleGuide={styleGuide}>
        <ApolloMockedProvider>
          <ThemeProvider theme={theme}>{children}</ThemeProvider>
        </ApolloMockedProvider>
      </NimbusConfig>
    ),
    ...options,
  });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
