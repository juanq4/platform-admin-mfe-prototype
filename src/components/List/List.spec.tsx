import { createSerializer, matchers } from "@emotion/jest";
import React from "react";
import { render, screen } from "../../utils/testUtils";
import { List } from "../List/List.component";

expect.addSnapshotSerializer(createSerializer());
expect.extend(matchers);

const mockHeader = "Mock header";
const mockFooter = "Mock footer";

const dataSourceToday = [
  { title: "today", content: "test 1 data" },
  { title: "today", content: "test 2 data" },
];

describe("Simple List Component", () => {
  test("7387811: [Given] no props are provided [Expect] the component to render", () => {
    const { container } = render(<List />);

    expect(container).toMatchSnapshot();
  });

  test("7387812: [Given] a header is provided [Expect] the header to show", () => {
    render(<List header={<h3>{mockHeader}</h3>} />);

    const header = screen.getByText(mockHeader);
    expect(header).toBeVisible();
  });

  test("7387813: [Given] a footer is provided [Expect] the footer to show", () => {
    render(<List footer={<h3>{mockFooter}</h3>} />);

    const footer = screen.getByText(mockFooter);
    expect(footer).toBeVisible();
  });

  test("7387814: [Given] data is provided [Expect] the data to show", () => {
    render(<List dataSource={dataSourceToday} renderItem={(item) => <div>{item.content}</div>} />);

    dataSourceToday.forEach((el) => {
      expect(screen.getByText(el.content)).toBeVisible();
    });

    const dataCount = screen.getAllByRole("listitem").length;
    expect(dataCount).toEqual(dataSourceToday.length);
  });

  test("7387815: [Given] data and a child exists [Expect] the child data to appear below the data", () => {
    const { container } = render(
      <List dataSource={dataSourceToday} renderItem={(item) => <div>{item.content}</div>}>
        <button>I am a button</button>
      </List>,
    );

    expect(container).toMatchSnapshot();
  });
});
