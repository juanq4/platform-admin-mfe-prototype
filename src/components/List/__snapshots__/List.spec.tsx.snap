// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Simple List Component 7387811: [Given] no props are provided [Expect] the component to render 1`] = `
.emotion-0 {
  list-style: none;
  margin: 0 0 20px 0;
  padding: 0;
}

.emotion-0 {
  list-style: none;
  margin: 0 0 20px 0;
  padding: 0;
}

<div>
  <ul
    class="emotion-0"
  />
</div>
`;

exports[`Simple List Component 7387815: [Given] data and a child exists [Expect] the child data to appear below the data 1`] = `
.emotion-0 {
  list-style: none;
  margin: 0 0 20px 0;
  padding: 0;
}

.emotion-1 {
  height: auto;
  font-size: 14px;
  line-height: 16px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-shadow: inset 0px -1px 0px inherit;
}

.emotion-0 {
  list-style: none;
  margin: 0 0 20px 0;
  padding: 0;
}

.emotion-1 {
  height: auto;
  font-size: 14px;
  line-height: 16px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-shadow: inset 0px -1px 0px inherit;
}

<div>
  <ul
    class="emotion-0"
  >
    <li
      class="emotion-1"
    >
      <div>
        test 1 data
      </div>
    </li>
    <li
      class="emotion-1"
    >
      <div>
        test 2 data
      </div>
    </li>
    <button>
      I am a button
    </button>
  </ul>
</div>
`;
